File: JavaScript/1382-Balance a Binary Search Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var balanceBST = function(root) {
    let nodeList = [];
    
    let traverseAdd = (node)=> {
        if (!node) return;
        traverseAdd(node.left);
        nodeList.push(node.val);
        traverseAdd(node.right);
    }
    traverseAdd(root);
    
    let construct = (left, right) =>{
        if (right < left) return null;
        let mid = left + Math.floor((right-left)/2);
        let currNode = new TreeNode(nodeList[mid]);
        currNode.left = construct(left, mid-1);
        currNode.right = construct(mid+1, right);
        return currNode
    }
    
    return construct(0, nodeList.length-1);
};

File: C++/1382-Balance a Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> allNodes;
    void dfs(TreeNode* node) {
        if (node == nullptr) return;
        dfs(node->left);
        allNodes.push_back(node->val);
        dfs(node->right);
    }
    TreeNode* construct(int left, int right) {
        if (left > right) return nullptr;
        if (left == right) return new TreeNode(allNodes[left]);
        int mid = left + (right-left)/2;

        return new TreeNode(allNodes[mid], construct(left, mid-1), construct(mid+1, right));
    }
    TreeNode* balanceBST(TreeNode* root) {
        dfs(root);
        return construct(0, allNodes.size()-1);
    }
};